{"cells": [{"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": "import hail as hl\nfrom hail.plot import output_notebook, show"}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": "import numpy as np\nimport pandas as pd"}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": "hl.init()\noutput_notebook()"}, {"cell_type": "markdown", "metadata": {}, "source": "#### Notebook to read in PVCF, filter to MAF > 1%, remove high missing sites and write out a bgen file"}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": "recode_contigs = {f\"{i}\":f\"chr{i}\" for i in (list(range(1, 23)) + [\"X\", \"Y\"])}"}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": "#read in file\npath_2_file = \"gs://pgr-wes-40k/GSA_chip_pvcf/CNCD_Freeze_Four.GT_hg38.pVCF.vcf.gz\"\n\ndata = hl.import_vcf(path_2_file, reference_genome='GRCh38', force_bgz=True, \\\n                     contig_recoding=recode_contigs, array_elements_required=False, skip_invalid_loci=True)"}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": "data.count()"}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": "data.rows().show()"}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": "#split multiallelic sites\nbi_allelics = data.filter_rows(hl.len(data.alleles) == 2)\nbi_allelics = bi_allelics.annotate_rows(a_index=1, was_split=False)\nmulti_allelics = data.filter_rows(hl.len(data.alleles) > 2)\nmulti_allelics_split = hl.split_multi_hts(multi_allelics)\ndata_split = multi_allelics_split.union_rows(bi_allelics)"}, {"cell_type": "code", "execution_count": null, "metadata": {"scrolled": true}, "outputs": [], "source": "#filter to biallelic sites only\n#missingness < 5%, aaf > 0.1%\n#excess heterozygosity pval > 1E-16\ndata_split = hl.variant_qc(data_split)\ndata_split_filtered = data_split.filter_rows((data_split.variant_qc.call_rate > 0.97))\nprint(data_split.count())\ndata_split_filtered = data_split_filtered.filter_rows((data_split_filtered.variant_qc.AF[1] > 0.01) & \\\n                                                (data_split_filtered.variant_qc.AF[1] < 0.99))\nprint(data_split_filtered.count())\ndata_split_filtered = data_split_filtered.filter_rows(data_split_filtered.variant_qc.p_value_excess_het > 1E-16)\nprint(data_split_filtered.count())"}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": "data_split_filtered = data_split_filtered\\\n.filter_rows(hl.is_snp(data_split_filtered.alleles[0], data_split_filtered.alleles[1]))\nprint(data_split_filtered.count())"}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": "'''\np = hl.plot.histogram(hl.log10(data\\\n                               .variant_qc.p_value_excess_het))\nshow(p)\n'''"}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": "bi_allelics_filtered\\\n.write(\"pvcf_MAF1_qced_split.mt\")"}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": "mt = hl.read_matrix_table(\"/vcf_MAF1_qced_split.mt\")\nmt_autosomes = mt.filter_rows((mt.locus.contig != \"chrX\"))\nmt_autosomes = mt_autosomes.filter_rows(mt_autosomes.locus.contig != \"chrY\")"}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": "mt_autosomes.count()"}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": "pruned_variant_table = hl.ld_prune(mt_autosomes.GT, r2=0.6, bp_window_size=75000, block_size=2048)\nld_pruned_mt = mt_autosomes.filter_rows(hl.is_defined(pruned_variant_table[mt_autosomes.row_key]))"}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": "print(ld_pruned_mt.count())"}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": "#Mean impute missing genotypes\nld_pruned_mt = ld_pruned_mt.annotate_rows(mean_gt = hl.agg.mean(ld_pruned_mt.GT.n_alt_alleles))\nld_pruned_mt = ld_pruned_mt.annotate_entries(n_alt_alleles=hl.coalesce(ld_pruned_mt.GT.n_alt_alleles, ld_prunded_mt.mean_gt))\n                                             \n#Add GP (genotype probability) field for BGEN file\nld_pruned_mt = ld_pruned_mt.annotate_entries(GP = GPs[ld_pruned_mt.GT.n_alt_alleles()])"}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": "hl.export_bgen(ld_pruned_mt, 'pvcf_MAF1_qced_split_autosomes.bgen')"}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ""}], "metadata": {"kernelspec": {"display_name": "Python 3", "language": "python", "name": "python3"}, "language_info": {"codemirror_mode": {"name": "ipython", "version": 3}, "file_extension": ".py", "mimetype": "text/x-python", "name": "python", "nbconvert_exporter": "python", "pygments_lexer": "ipython3", "version": "3.6.13"}}, "nbformat": 4, "nbformat_minor": 5}